sklearn/tree/tests/test_monotonic_tree.py

```diff
@@ -0,0 +1,235 @@
+import pytest
+from sklearn import datasets
+from sklearn.tree import DecisionTreeRegressor
+from sklearn.tree.tests.test_tree import REG_TREES, CLF_TREES
+
+
+@pytest.mark.parametrize('seed', range(4))
+@pytest.mark.parametrize('depth_first', (True, False))
+def test_montonic_constraints(seed, depth_first):
+    n_samples = 1000
+    n_samples_train = 900
+    X, y = datasets.make_hastie_10_2(n_samples=n_samples, random_state=0)
+    train = np.arange(n_samples_train)
+    test = np.arange(n_samples_train, n_samples)
+    X_train = X[train]
+    y_train = y[train]
+    X_test = np.copy(X[test])
+
+    X_test_incr = np.copy(X_test)
+    X_test_decr = np.copy(X_test)
+    X_test_incr[:, 0] += 10
+    X_test_decr[:, 1] += 10
+    monotonic_cst = np.zeros(X.shape[1])
+    monotonic_cst[0] = 1
+    monotonic_cst[1] = -1
+
+    for name, TreeRegressor in REG_TREES.items():
+        if depth_first:
+            est = TreeRegressor(max_depth=None, monotonic_cst=monotonic_cst)
+        else:
+            est = TreeRegressor(max_depth=None, monotonic_cst=monotonic_cst,
+                                max_leaf_nodes=n_samples_train)
+        if hasattr(est, "random_state"):
+            est.set_params(**{"random_state": seed})
+        est.fit(X_train, y_train)
+        y = est.predict(X_test)
+        # increasing constraint
+        y_incr = est.predict(X_test_incr)
+        # y_incr should always be greater than y
+        assert np.all(y_incr >= y)
+
+        # decreasing constraint
+        y_decr = est.predict(X_test_decr)
+        # y_decr should always be lower than y
+        assert np.all(y_decr <= y)
+
+    for name, TreeClassifier in CLF_TREES.items():
+        if depth_first:
+            est = TreeClassifier(max_depth=None, monotonic_cst=monotonic_cst)
+        else:
+            est = TreeClassifier(max_depth=None, monotonic_cst=monotonic_cst,
+                                 max_leaf_nodes=n_samples_train)
+        if hasattr(est, "random_state"):
+            est.set_params(**{"random_state": seed})
+        est.fit(X_train, y_train)
+        y = est.predict_proba(X_test)[:, 0]
+
+        # increasing constraint
+        y_incr = est.predict_proba(X_test_incr)[:, 0]
+        # y_incr should always be greater than y
+        assert np.all(y_incr >= y)
+
+        # decreasing constraint
+        y_decr = est.predict_proba(X_test_decr)[:, 0]
+        # y_decr should always be lower than y
+        assert np.all(y_decr <= y)
+
+
+def test_multiclass_raises():
+    X, y = datasets.make_hastie_10_2(n_samples=100, random_state=0)
+    y[0] = 0
+    monotonic_cst = np.zeros(X.shape[1])
+    monotonic_cst[0] = -1
+    monotonic_cst[1] = 1
+    for name, TreeClassifier in CLF_TREES.items():
+        est = TreeClassifier(max_depth=None, monotonic_cst=monotonic_cst,
+                             random_state=0)
+        if hasattr(est, "random_state"):
+            est.set_params(**{"random_state": 0})
+
+        with pytest.raises(ValueError):
```

Comment:
Could you please also check that the error message is informative with the `match` kwarg of `pytest.raises`?

Reply:
Sure !